text/0002-extensible-provisioning-templates.md

```diff
@@ -0,0 +1,118 @@
+
+1.  Foreman will generate a certificate during the regular orchestration process and store it as part of
+    host's information
+2.  Finish template will be extended by the puppet plugin so it will install the certificate to the newly
+    created host.
+
+This method has many advantages that are out of scope for this RFC.
+
+## Advantages of extensible templates
+
+1.  Encapsulation of plugin related template code inside the plugin: no more templates that are not compatible
+    with certain versions of a plugin.
+2.  Providing a valid extension point for actions that should be executed upon actual host build.
+
+# Detailed design
+
+[design]: #detailed-design
+
+We will use existing mechanism for extending views called [pagelets](http://projects.theforeman.org/projects/foreman/wiki/Pagelets).
+We will add a call to render pagelets that are already registered each time a template is rendered and pass all the information about the template.
+Any plugin will be able to register its own partial that will be embedded into the final template.
+
+## Changes to the core
+
+1.  Add a call to pagelets to the template sent for rendering in `UnattendedController#safe_render`:
+
+```ruby
+render :inline => "<%= unattended_render(@unsafe_template, @template_name).html_safe %> <%= render_pagelets_for(:template_extension, :subject => @template_object) %>" and return
+```
+
+2.  Remove `exit 0` line from "Kikstart finish template"
+3.  (optional) Register a pagelet extension with high priority to add the `exit 0` line after all plugins have rendered.
+
+Registration:
+
+```ruby
+Pagelets::Manager.add_pagelet(
+  "unattended/host_template",
+  :template_extension,
+  :partial => 'unattended/kickstart/finish/after_plugins',
+  :priority => 2000,
+  :onlyif => ->(subject){ subject.name == 'Kickstart default finish' })
+```
+
+## Usage
+
+1.  Define a partial template that should be rendered.
+2.  Register the template by calling the `#add_pagelet` method
+3.  Don't forget to add a condition that will filter the partial rendering to a specific templates.
+
+Example:
+
+`unattended/kickstart/finish/networking.erb`:
+
+    real=`ip -o link | grep 52:54:00:8c:81:40 | awk '{print $2;}' | sed s/:$//`
+
+    cat << EOF > /etc/sysconfig/network-scripts/ifcfg-$real
+    BOOTPROTO="none"
+    IPADDR=""
+    NETMASK="255.255.255.0"
+    DEVICE="$real"
+    HWADDR="52:54:00:8c:81:40"
+    ONBOOT=yes
+    PEERDNS=yes
+    PEERROUTES=yes
+    EOF
+
+    service network restart
+
+registration:
+
+```ruby
+Pagelets::Manager.add_pagelet(
+  "unattended/host_template",
+  :template_extension,
+  :partial => 'unattended/kickstart/finish/networking',
+  :priority => 100,
+  :onlyif => ->(subject){ subject.name == 'Kickstart default finish' }) # <- the partial will be rendered only for this template
+```
+
+# Alternatives
+
+[alternatives]: #alternatives
+
+Main alternative will be leaving `#handle_ca` in core and make it extensible. IMHO it will only add to Foreman's technical
```

Comment:
What I like about this proposal is splitting up things that are genuinely orchestration (e.g. generating the cert on the puppet master) from the task of transporting the secrets to the host.

Today we hit issues where a template will render fine in preview/spoof mode but fail when actually used because the proxy doesn't have (for example) sudo rights to call puppet cert.

With this proposal, we can move cert generation to the puppet orchestration queue (because that's not autosign, it's still safe) which can fail gracefully just like any DNS or DHCP error - and then we can deal with transporting that cert at build-time.